velox/common/base/RawVector.h

```diff
@@ -177,7 +183,8 @@ class raw_vector {
 
   void freeData(T* data) {
     if (data_) {
-      ::free(addBytes(data, -simd::kPadding));
+      auto bytes = paddedSize(sizeof(T) * allocatedSize_);
```

Comment:
`allocatedSize_` is already padded byte size, just pass it directly to `deallocate`

Reply:
Yes, this was actually a bug. But it revealed a problem with the interfaces of the two allocators used.

The original implementation used byte size. But the AlignedStlAllocator (HashStringAllocator) wants as input the number of elements to do its own padding/alignment.
The problem is that the result isn't a memory block that neatly multiples of sizeof(T) and won't necessarily align with the xsimd alignment that is used for the raw vectors.
I'm seeing SEGVs in the fuzzers because the raw vector for the hashes vector during simd execution appears to be reading the incorrect number of rows. My guess is that the memory layout due to the incorrect simd alignment of the rows vector causes this issue.

Reply:
There is also a bug in `StdAlignedAllocator::allocate`:
```
    return reinterpret_cast<T*>(aligned_alloc(kAlignment, n * sizeof(T)));
```

Reply:
Right. I did fix that when I noticed it is the number of elements.
There is some kind of memory corruption with the memory from the HSA. I can see that the rows vector would contain sequential numbers (the row numbers) but I can see it being corrupted when it is freed or when it is occasionally read. Any contents in the memory region that was allocated from the HSA. I've seen rows values being corrupted, and in some cases it is the delta that is used to calculate the actual pointer that was allocated, sometimes this corrupted value is accessed to find the hash in the probe and fails, the HSA header could also be corrupted, etc. And it contains non-rows content - clearly non-numeric values (some values seem to be ascii strings - e.g. `E_ARRAY` and others).
I'm trying to figure out what is happening. The memory block allocated previously outside the HSA didn't have this kind of problem.

Reply:
Are you using the `raw_vector` outside the life cycle of HSA?

Reply:
I don't think so. Say the RowNumber operator that is used for one of the failing fuzzers. The lookup_ is destructed first and the table_ with the HSA after that. So this should be ok. If HSA was invalid we would get SEGV on access of the allocate/deallocate during deallocation of the raw_vector in the HashLookup. 
The issue doesn't always happen.

Reply:
I've found the root cause. The issue is that we pass the HSA as a regular point while it is a shared_ptr in the RowContainer that has only a single reference. Later when restoring the spill partition data, the hash table is cleared, which also clears the RowContainer  (`RowContainer::clear()`) and if the the HSA is not shared then it also clears out the HSA and with it the raw_vector data in the HashLookup leading to subsequent lookup problems.

I see two possibilities:
1. Let each opertor using the HashLookup use its operator pool to create a new HSA for it. This affects the HashProbe, RowNumber, TopNRowNumber operators. And also have the HashTable use its own memory pool creating the HashLookup object whenevr needed. So they are separate from each other.
2. Keep a shared_ptr reference to the RowContainer HSA in the HashLookup to make sure the HSA is shared (use count will be 2) is, therefore, not cleared.
3. 
Is there a preference? Option 2 would likely involve much less allocation/deallocation of a new HSA if the HashLookup is frequently destroyed/created.

Reply:
Option 2 is probably not possible because we must clear out and release the memory when spilling is triggered.  If we want to continue to use the same HSA as row container, all these `raw_vector`s should also be spilled and unspilled at the same time as the row container.  @xiaoxmeng Do you think this is doable?

Option 1 is possible but might have performance issue.  Maybe we can have a driver level HSA to be used here?